ext/simdjson/src/bindings.cpp

```diff
@@ -0,0 +1,238 @@
+  +----------------------------------------------------------------------+
+  | This source file is subject to version 2.0 of the Apache license,    |
+  | that is bundled with this package in the file LICENSE, and is        |
+  | available through the world-wide-web at the following url:           |
+  | http://www.apache.org/licenses/LICENSE-2.0.html                      |
+  +----------------------------------------------------------------------+
+  | Author: Jinxi Wang  <1054636713@qq.com>                              |
+  | Author: Máté Kocsis <kocsismate@woohoolabs.com>                      |
+  +----------------------------------------------------------------------+
+*/
+
+#include "php.h"
+#include "php_simdjson.h"
+
+#include "simdjson.h"
+#include "bindings.h"
+
+extern "C" {
+	#include <ext/json/php_json.h>
+	#include <Zend/zend_exceptions.h>
+}
+
+// see https://github.com/simdjson/simdjson/blob/master/doc/performance.md#reusing-the-parser-for-maximum-efficiency
+simdjson::dom::parser parser;
+
+WARN_UNUSED
+simdjson::error_code
+build_parsed_json_cust(
+	simdjson::dom::element &doc, const char *buf, size_t len, bool realloc_if_needed, u_short depth = simdjson::DEFAULT_MAX_DEPTH
+) {
+    auto error = parser.allocate(len, depth);
+
+    if (error) {
+        return error;
+    }
+
+    error = parser.parse(buf, len, realloc_if_needed).get(doc);
+    if (error) {
+        return error;
+    }
+
+    return simdjson::SUCCESS;
+}
+
+static zval create_array(simdjson::dom::element element)
+{
+    zval v;
+
+    switch (element.type()) {
+        //ASCII sort
+        case simdjson::dom::element_type::STRING :
+            ZVAL_STRING(&v, std::string_view(element).data());
+            break;
+        case simdjson::dom::element_type::INT64 : ZVAL_LONG(&v, int64_t(element));
+            break;
+        case simdjson::dom::element_type::UINT64 : ZVAL_LONG(&v, uint64_t(element));
+            break;
+        case simdjson::dom::element_type::DOUBLE : ZVAL_DOUBLE(&v, double(element));
+            break;
+        case simdjson::dom::element_type::BOOL :
+            ZVAL_BOOL(&v, bool(element));
+            break;
+        case simdjson::dom::element_type::NULL_VALUE :
+            ZVAL_NULL(&v);
+            break;
+        case simdjson::dom::element_type::ARRAY :
+            zval arr;
+            array_init(&arr);
+
+            for (simdjson::dom::element child : simdjson::dom::array(element)) {
+                zval value = create_array(child);
+                add_next_index_zval(&arr, &value);
+            }
+
+            v = arr;
+            break;
+        case simdjson::dom::element_type::OBJECT :
+            zval obj;
+            array_init(&obj);
+
+            for (simdjson::dom::key_value_pair field : simdjson::dom::object(element)) {
+                zval value = create_array(field.value);
+                add_assoc_zval_ex(&obj, field.key.data(), strlen(field.key.data()), &value);
```

Comment:
```c
  /**
   * Cast this element to a null-terminated C string.
   *
   * The string is guaranteed to be valid UTF-8.
   *
   * The get_c_str() function is equivalent to get<const char *>().
   *
   * The length of the string is given by get_string_length(). Because JSON strings
   * may contain null characters, it may be incorrect to use strlen to determine the
   * string length.
   *
   * It is possible to get a single string_view instance which represents both the string
   * content and its length: see get_string().
   *
   * @returns A pointer to a null-terminated UTF-8 string. This string is stored in the parser and will
   *          be invalidated the next time it parses a document or when it is destroyed.
   *          Returns INCORRECT_TYPE if the JSON element is not a string.
   */
  inline simdjson_result<const char *> get_c_str() const noexcept;
  /**
   * Gives the length in bytes of the string.
   *
   * It is possible to get a single string_view instance which represents both the string
   * content and its length: see get_string().
   *
   * @returns A string length in bytes.
   *          Returns INCORRECT_TYPE if the JSON element is not a string.
   */
  inline simdjson_result<size_t> get_string_length() const noexcept;

```

Instead of data(), it seems like it may be more efficient to use get_c_str() and get_string_length()

Reply:
I've been trying to find out how I could invoke `key_c_str` on the iterator (https://simdjson.org/api/0.4.1/classsimdjson_1_1dom_1_1object_1_1iterator.html#a54a99dcf9c422d5c24f3c570b6d9b6e4). Do you have any idea?

Reply:
I think you would want to iterate over simdjson::dom::object::iterator() instead of the key_value_pair that it provides a helper to cast to.

```
/**
 * JSON object.
 */
class object {
public:
  /** Create a new, invalid object */
  really_inline object() noexcept;

  class iterator {
  public:
    /**
     * Get the actual key/value pair
     */
    inline const key_value_pair operator*() const noexcept;
```

You can iterate manually over begin() and end() `for (simdjson::dom::object::iterator it = begin(); it != end(); it++)`

Reply:
Works! I was trying to achieve something similar, but I wasn't aware of this possibility :)